### PR TITLE
Bug/prism/483/form-data-validation

### PR DESCRIPTION
**Pull request by chris-miaskowski**
https://github.com/stoplightio/prism/issues/483
 https://github.com/stoplightio/prism/pull/482

## What was the issue

Form Data in OAS2 is much different than in OAS3. E.g. in OAS3 everything is a schema. In OAS2 on the other hand form data is a parameter similar to path or query param. We tranformed OAS2 form data parameter incorrectly by throwing everything in `encoding`. This way the `schema` was incorrecly built and Prism could validate it. 

## Changes

- refactor body form-data validation
- update tests

### CODE REVIEW

File: src/oas2/transformers/params.ts [hunk 1]
```diff
@@ -81,41 +82,84 @@ export function translateToBodyParameter(body: BodyParameter, consumes: string[]
 }
 
 export function translateFromFormDataParameter(
-  form: FormDataParameter,
-  requestBody: IHttpOperationRequestBody | null | undefined,
+  parameters: FormDataParameter[],
   consumes: string[],
 ): IHttpOperationRequestBody {
-  const baseContent = buildContentFromParameter(form);
```
Comment (XVincentX) [inline src/oas2/transformers/params.ts]: Apparently you've changed this function quite a lot. According to the code this was transforming a *single* parameter, now it's transforming all the parameters all together. Was there any particular reason for this?

File: src/oas2/transformers/params.ts [hunk 2]
```diff
@@ -81,41 +82,84 @@ export function translateToBodyParameter(body: BodyParameter, consumes: string[]
...
+          content.encodings.push(encoding);
+        }
+      }
+    });
+    return body;
+  }, finalBody);
+}
 
-    bodyContent.encodings = (bodyContent.encodings || []).concat(encoding);
+function buildEncoding(parameter: FormDataParameter, content: IMediaTypeContent): IHttpEncoding | null {
```
Comment (XVincentX) [inline src/oas2/transformers/params.ts]: The `content` property seems not used, can we remove it?

File: src/oas2/transformers/params.ts [hunk 3]
```diff
@@ -157,11 +201,9 @@ function buildContentFromParameter(
...
   );
 
-  return {
-    schema: pickBy(
-      defaults<Schema, Schema>(schema, {
-        minLength: 'allowEmptyValue' in param && param.allowEmptyValue ? 1 : undefined,
-      }),
-    ),
-  };
+  if ('allowEmptyValue' in param && param.allowEmptyValue) {
```
Comment (XVincentX) [inline src/oas2/transformers/params.ts]: Yeah, I admit the readability of this line of code is way better than the previos one 👍

File: src/oas2/transformers/params.ts [hunk 4]
```diff
@@ -81,41 +82,84 @@ export function translateToBodyParameter(body: BodyParameter, consumes: string[]
...
-      property: form.name,
-      style: HttpParamStyles.Form,
-    };
+  return parameters.reduce((body, parameter) => {
+    const { schema } = buildSchemaForParameter(parameter);
+    (body.contents || []).forEach(content => {
+      // workaround... JSONSchema4 does not support `allowEmptyValue`
+      if ('allowEmptyValue' in parameter) {
+        // @ts-ignore
+        schema.allowEmptyValue = parameter.allowEmptyValue;
```
Comment (XVincentX) [inline src/oas2/transformers/params.ts]: You can use `Object.assign` to avoid to shutdown the type checker on this line

```suggestion
        Object.assign(schema, { allowEmptyValue: parameter.allowEmptyValue });
```

File: src/oas2/transformers/params.ts [hunk 5]
```diff
@@ -81,41 +82,84 @@ export function translateToBodyParameter(body: BodyParameter, consumes: string[]
...
+      }
+      set(content, `schema.properties.${parameter.name}`, schema);
+      if (parameter.required) {
+        const requiredIndex = get(content, 'schema.required', []).length;
```
Comment (XVincentX) [inline src/oas2/transformers/params.ts]:
```suggestion
        const requiredIndex = get(content, 'schema.required.length', 0);
```

File: src/oas2/transformers/request.ts [hunk 6]
```diff
@@ -11,21 +11,31 @@ import {
...
-      const headerParameter = translateToHeaderParam(parameter);
-      request.headers = (request.headers || []).concat(headerParameter);
-    }
-    return request;
-  }, {});
+  const bodyParameters = parameters.filter(parameter => isBodyParameter(parameter)) as BodyParameter[];
+  const formDataParameters = parameters.filter(parameter => isFormDataParameter(parameter)) as FormDataParameter[];
+  const request: IHttpOperationRequest = {};
+
+  // if 'body' and 'form data' defined prefer 'body'
```
Comment (XVincentX) [inline src/oas2/transformers/request.ts]: Yeah I think this is a good choice. The spec does not really mandate that a parameter with `formData` excludes `body` and viceversa.

File: src/oas2/transformers/params.ts [hunk 7]
```diff
@@ -19,7 +20,7 @@ import {
 } from 'swagger-schema-official';
 
 function chooseQueryParameterStyle(
```
Comment (XVincentX) [inline src/oas2/transformers/params.ts]: This function is not taking a `FormDataParameter` as well — so the name is probably not ok anymore; in any case this is only used in the `QueryParameter` path so I'm not too sure why this change is here

File: src/oas2/transformers/params.ts (hunk 1 above)
Comment (chris-miaskowski) [reply src/oas2/transformers/params.ts]: Hey, good question. Several reasons:
1. The original function horrible - it "kind of mutated inputs"...
2. An array of parameters is reduced into a single schema - it previously took each param individually and "existing body" in order to "mutate" that existing body. That was redundant and confusing.
3. It didn't build `encodings` correctly.

File: src/oas2/transformers/params.ts [hunk 8]
```diff
@@ -81,41 +82,84 @@ export function translateToBodyParameter(body: BodyParameter, consumes: string[]
...
+        style: HttpParamStyles.Form,
+        explode: true,
+      };
+    case 'ssv':
+      return {
+        property: parameter.name,
+        style: HttpParamStyles.SpaceDelimited,
+        explode: false,
+      };
+    // TODO: TSV is not supported in OAS3, no compatibility
```
[the remark this answers is not in the document]
Comment (chris-miaskowski) [reply src/oas2/transformers/params.ts]: Will remove the comment and create an issue. The point I was trying to make is to leave a reminder that our current types do not allow for TSV. We probably need to make our types more flexible.

File: src/oas2/transformers/request.ts [hunk 9]
```diff
@@ -11,21 +11,31 @@ import {
...
-    } else if (isPathParameter(parameter)) {
-      const pathParameter = translateToPathParameter(parameter);
-      request.path = (request.path || []).concat(pathParameter);
-    } else if (isHeaderParameter(parameter)) {
-      const headerParameter = translateToHeaderParam(parameter);
-      request.headers = (request.headers || []).concat(headerParameter);
-    }
-    return request;
-  }, {});
+  const bodyParameters = parameters.filter(parameter => isBodyParameter(parameter)) as BodyParameter[];
```
[the remark this answers is not in the document]
Comment (chris-miaskowski) [reply src/oas2/transformers/request.ts]: Yup, when passed function directly it works like a charm.

File: src/oas2/transformers/params.ts (hunk 7 above)
Comment (chris-miaskowski) [reply src/oas2/transformers/params.ts]: Oh, overshight! Removing!

File: src/oas2/transformers/params.ts (hunk 1 above)
Comment (XVincentX) [reply src/oas2/transformers/params.ts]: In that case we should change the name
